onnxruntime/core/graph/graph_flatbuffers_utils.cc

```diff
@@ -339,4 +340,134 @@ Status LoadAttributeOrtFormat(const fbs::Attribute& fbs_attr,
+    tb.add_raw_data(raw_data);
+    fbs_tensor = tb.Finish();
+    return Status::OK();
+  };
+
+  const onnxruntime::Tensor& src_tensor = ort_value.Get<onnxruntime::Tensor>();
+
+  // Check if the tensor is on CPU. If not, we need to copy the tensor to CPU before saving it.
+  if (const auto& tensor_location = src_tensor.Location();
+      tensor_location.device.Type() != OrtDevice::CPU) {
+    InlinedVector<uint8_t> tensor_data_buffer{};
+    tensor_data_buffer.resize(src_tensor.SizeInBytes());
+    const OrtMemoryInfo cpu_alloc_info{onnxruntime::CPU, OrtDeviceAllocator};
+    onnxruntime::Tensor dst_tensor{src_tensor.DataType(), src_tensor.Shape(), tensor_data_buffer.data(), cpu_alloc_info};
+    ORT_RETURN_IF_ERROR(copy_tensor(src_tensor, dst_tensor));
+    ORT_RETURN_IF_ERROR(ort_tensor_to_fbs_tensor(tensor_name, dst_tensor, fbs_tensor));
+  } else {
+    ORT_RETURN_IF_ERROR(ort_tensor_to_fbs_tensor(tensor_name, src_tensor, fbs_tensor));
+  }
+
+  return Status::OK();
+}
+
+#define UNPACK_TESOR_WITH_TYPE(T)                                                                          \
+  ORT_RETURN_IF_ERROR(onnxruntime::utils::UnpackTensor(unused_tensor_proto, fbs_tensor.raw_data()->Data(), \
+                                                       fbs_tensor.raw_data()->size(),                      \
+                                                       dst_tensor->MutableData<T>(),                       \
+                                                       static_cast<size_t>(dst_tensor->Shape().Size())));  \
+  break;
+
+Status LoadOrtValueOrtFormat(const fbs::Tensor& fbs_tensor,
+                             std::string& tensor_name, OrtValue& ort_value) {
+  // The assumption is that the flatbuffer buffer will be destructed once the checkpoint has been loaded.
+  // And so, we must allocate a buffer where the tensor data can be copied using the cpu allocator.
+  // This buffer is owned by the OrtValue.
+  static const CPUExecutionProviderInfo info;
+  static const CPUExecutionProvider cpu_provider(info);
+  static const AllocatorPtr cpu_allocator = cpu_provider.GetAllocator(OrtMemTypeDefault);
+
+  auto* fbs_tensor_name = fbs_tensor.name();
+  ORT_RETURN_IF_NOT(fbs_tensor_name, "Checkpoint is invalid. Expected: A valid tensor name. Actual: nullptr.");
+  tensor_name = fbs_tensor_name->str();
+
+  auto* tensor_dims = fbs_tensor.dims();
+  ORT_RETURN_IF_NOT(tensor_dims, "Checkpoint is invalid. Expected: Valid tensor dims. Actual: nullptr.");
+
+  auto tensor_data_type = fbs_tensor.data_type();
+  const DataTypeImpl* tensor_dtype = DataTypeImpl::TensorTypeFromONNXEnum(
+                                         static_cast<int>(tensor_data_type))
+                                         ->GetElementType();
+  auto dst_tensor = std::make_unique<onnxruntime::Tensor>(
+      tensor_dtype, TensorShape(InlinedVector<int64_t>{tensor_dims->begin(), tensor_dims->end()}), cpu_allocator);
```

Comment:
instead of the intermediate InlinedVector, can we just call the TensorShape ctor that takes a pointer and size?
https://github.com/microsoft/onnxruntime/blob/fa828ed086d937293b289255978ddfc0f183faf8/include/onnxruntime/core/framework/tensor_shape.h#L87